File: C++/Bryann Valderrama/Strings/finiteAutomata.cpp
#include <cstdlib>
#include <bits/stdc++.h>
using namespace std;
#define NO_OF_CHARS 256

int getNextState(char pat[], int M, int state, int x)
{
    if (state < M && x == pat[state])
    {
        return state + 1;
    }
    int ns, i;
    for (ns = state; ns > 0; ns--)
    {
        if (pat[ns - 1] == x)
        {
            for (i = 0; i < ns - 1; i++)
            {
                if (pat[i] != pat[state - ns + 1 + i])
                {
                    break;
                }
            }
            if (i == ns - 1)
            {
                return ns;
            }
        }
    }
    return 0;
}

void computeTF(char pat[], int M, int TF[NO_OF_CHARS][NO_OF_CHARS])
{
    int state, x;
    for (state = 0; state <= M; state++)
    {
        for (x = 0; x < NO_OF_CHARS; x++)
        {
            TF[state][x] = getNextState(pat, M, state, x);
        }
    }
}

void finiteAutomata(char pat[], char txt[])
{
    int M = strlen(pat);
    int N = strlen(txt);
    int TF[M + 1][NO_OF_CHARS];
    computeTF(pat, M, TF);
    int i, state = 0;
    for (i = 0; i < N; i++)
    {
        state = TF[state][txt[i]];
        if (state == M)
        {
            cout << "Patron encontrado en indice " << i - M + 1 << " - " << i << endl;
        }
    }
}

int main(int argc, char const *argv[])
{
    char pat[] = {"looloolo"};
    char txt[] = {"olo"};
    finiteAutomata(txt, pat);
    return 0;
}

File: Codigos estudiantes por lenguaje/PY/Bryann Valderrama/Strings/finiteAutomata.py
from sys import stdin, stdout
NO_OF_CHARS = 256


def getNextState(pat, M, state, x):
    if state < M and x == ord(pat[state]):
        return state + 1
    i = int(0)
    for ns in range(state, 0, -1):
        if ord(pat[ns-1]) == x:
            for i in range(ns-1):
                if pat[i] != pat[state-ns+1+i]:
                    break
                i += 1
            if i == ns-1:
                return ns
    return 0


def computeTF(pat, M, TF):
    global NO_OF_CHARS
    for state in range(M+1):
        for x in range(NO_OF_CHARS):
            TF[state][x] = getNextState(pat, M, state, x)


def finiteAutomata(pat, txt):
    global NO_OF_CHARS
    M = len(pat)
    N = len(txt)
    TF = [[None for x in range(NO_OF_CHARS)] for x in range(M+1)]
    computeTF(pat, M, TF)
    state = int(0)
    for i in range(N):
        state = TF[state][ord(txt[i])]
        if state == M:
            stdout.write(f'Patron encontrado en indice {i-M+1} - {i}\n')


txt = [x for x in 'looloolo']
pat = [x for x in 'olo']
finiteAutomata(pat, txt)
